### PR TITLE
[SPARK-13629] [ML] Add binary toggle Param to CountVectorizer

### PR DESCRIPTION
**Pull request by hhbyyh**
## What changes were proposed in this pull request?

It would be handy to add a binary toggle Param to CountVectorizer, as in the scikit-learn one: http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html
If set, then all non-zero counts will be set to 1.
## How was this patch tested?

unit tests

### CODE REVIEW

File: mllib/src/test/scala/org/apache/spark/ml/feature/CountVectorizerSuite.scala [hunk 1]
```diff
@@ -168,6 +168,24 @@ class CountVectorizerSuite extends SparkFunSuite with MLlibTestSparkContext
...
   }
 
+  test("CountVectorizerModel with binary") {
+    val df = sqlContext.createDataFrame(Seq(
+      (0, split("a a a b b c"), Vectors.sparse(4, Seq((0, 1.0), (1, 1.0), (2, 1.0)))),
+      (1, split("c c c"), Vectors.sparse(4, Seq((2, 1.0)))),
+      (2, split("a"), Vectors.sparse(4, Seq((0, 1.0))))
+    )).toDF("id", "words", "expected")
+
+    // minTF: count
```
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/feature/CountVectorizerSuite.scala]: Looks like this comment was copied from above test case. Not applicable.

Comment (hhbyyh) [inline mllib/src/test/scala/org/apache/spark/ml/feature/CountVectorizerSuite.scala]: Thanks @MLnick .

**Comment by MLnick** [comment]
@hhbyyh looks fine to me - can you take a look at how easy it will be to add similar functionality to `HashingTF`? If it's easy, you can add that to this PR and update the JIRA. If not, I'll create a separate JIRA for it.

**Comment by MLnick** [comment]
jenkins retest this please

**Comment by hhbyyh** [comment]
Sorry for the late response. @MLnick and thanks for your review. I can check for HashingTF tomorrow if you prefer the merge the changes together.

**Comment by MLnick** [comment]
@hhbyyh I'm going to merge this and `HashingTF` can be done separately (it looks trivial from my quick glance)

**Comment by hhbyyh** [comment]
Thanks. Let me know if I can help in any way.

**Comment by MLnick** [comment]
Merged into master. Thanks! I've created a ticket for `HashingTF` [here](https://issues.apache.org/jira/browse/SPARK-13963) and for PySpark [here](https://issues.apache.org/jira/browse/SPARK-13967)

**Comment by jkbradley** [comment]
@hhbyyh Thanks for the PR!  LGTM, I agree with the use of minTF

**Comment by jkbradley** [comment]
Oh, actually, I have a minor comment.  Could you please send a cleanup PR for it?

File: mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala [hunk 2]
```diff
@@ -207,6 +207,27 @@ class CountVectorizerModel(override val uid: String, val vocabulary: Array[Strin
   /** @group setParam */
   def setMinTF(value: Double): this.type = set(minTF, value)
 
+  /**
+    * Binary toggle to control the output vector values.
+    * If True, all non zero counts are set to 1. This is useful for discrete probabilistic
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala]: Could you clarify that this ("all non-zero counts") is after the minTF filter has been applied?
(Also, as long as you're at it, could you please correct the indentation?)

File: mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala [hunk 3]
```diff
@@ -233,7 +254,13 @@ class CountVectorizerModel(override val uid: String, val vocabulary: Array[Strin
       } else {
         tokenCount * minTf
       }
-      Vectors.sparse(dictBr.value.size, termCounts.filter(_._2 >= effectiveMinTF).toSeq)
+      val effectiveCounts = if ($(binary)) {
+        termCounts.filter(_._2 >= effectiveMinTF).map(p => (p._1, 1.0)).toSeq
+      }
+      else {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala]: Combine with previous line

**Comment by jkbradley** [comment]
That's it--thanks!

**Comment by hhbyyh** [comment]
Get it. Thanks.

**Pull request closed without merging**